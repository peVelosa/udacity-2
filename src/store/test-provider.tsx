import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { RootState, AppStore, setupStore } from ".";
import { MemoryRouter, Routes, Route, InitialEntry } from "react-router";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {},
  initialEntries: InitialEntry[],
  path: string
) {
  const {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>
      <MemoryRouter initialEntries={initialEntries}>
        <Routes>
          <Route path={path} element={children} />
        </Routes>
      </MemoryRouter>
      ,
    </Provider>
  );

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
