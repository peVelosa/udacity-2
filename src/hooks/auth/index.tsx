import { useAppSelector } from "@/store/hooks";

export const useAuth = () => {
  const { isAuthenticated } = useAppSelector((state) => state.auth);

  return {
    isAuthenticated,
  };
};
