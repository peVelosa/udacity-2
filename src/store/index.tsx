import { combineReducers, configureStore } from "@reduxjs/toolkit";
import { authSlice } from "./auth/auth.reducer";
import { poolSlice } from "./pool/pool.reducer";
import { usersSlice } from "./users/users.reducer";

export const rootReducer = combineReducers({
  auth: authSlice.reducer,
  questions: poolSlice.reducer,
  users: usersSlice.reducer,
});

export function setupStore(preloadedState?: Partial<RootState>) {
  return configureStore({
    reducer: rootReducer,
    preloadedState,
  });
}
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore["dispatch"];
