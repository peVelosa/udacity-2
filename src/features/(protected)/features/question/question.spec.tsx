import "@testing-library/jest-dom";
import { QuestionPage } from "./";
import { renderWithProviders } from "@/store/test-provider";
import { MemoryRouter, Route, Routes } from "react-router";
import * as redux from "@/store/hooks";

const preloadedState = {
  auth: {
    user: {
      id: "user1",
      password: "password", // dummy value
      name: "Test User",
      avatarURL: "http://example.com/avatar.png",
      answers: {},
      questions: [],
    },
    isAuthenticated: true,
  },
  questions: {
    questions: [
      {
        id: "q1",
        author: "user1",
        timestamp: 1467166872634,
        optionOne: {
          votes: ["user1", "user2"], // 2 votes
          text: "Option One text",
        },
        optionTwo: {
          votes: ["user3", "user4", "user5"], // 3 votes
          text: "Option Two text",
        },
      },
    ],
    error: null,
    isLoading: false,
    selectedQuestion: {
      author: {
        id: "user1",
        password: "password", // dummy value
        name: "Test User",
        avatarURL: "http://example.com/avatar.png",
        answers: {},
        questions: [],
      },
      question: {
        id: "q1",
        author: "user1",
        timestamp: 1467166872634,
        optionOne: {
          votes: ["user1", "user2"],
          text: "Option One text",
        },
        optionTwo: {
          votes: ["user3", "user4", "user5"],
          text: "Option Two text",
        },
      },
    },
  },
} as const;

jest.mock("util", () => ({
  TextEncoder: class {},
  TextDecoder: class {},
}));

describe("QuestionPage Vote Percentages", () => {
  test("displays correct percentages for each option and dispatches actions", () => {
    jest.spyOn(redux, "useAppDispatch").mockImplementation(() => jest.fn());

    const { getByTestId } = renderWithProviders(
      <MemoryRouter initialEntries={["/questions/q1"]}>
        <Routes>
          <Route path="/questions/:questionId" element={<QuestionPage />} />
        </Routes>
      </MemoryRouter>,
      { preloadedState: preloadedState as any }
    );

    const optionOne = getByTestId("total-option-one");
    const optionTwo = getByTestId("total-option-two");

    expect(optionOne).toHaveTextContent("40");
    expect(optionTwo).toHaveTextContent("60");
  });
});
