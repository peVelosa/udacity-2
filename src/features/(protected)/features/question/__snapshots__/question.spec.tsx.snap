// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionPage Vote Percentages matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="space-y-8"
  >
    <h1
      class="text-xl font-bold text-center"
    >
      Poll by Test User
    </h1>
    <img
      alt="Test User"
      class="w-16 h-16 rounded-full mx-auto"
      src="http://example.com/avatar.png"
    />
    <h2
      class="text-xl font-bold text-center"
    >
      Would you rather
    </h2>
    <div
      class="flex items-center gap-4 justify-center"
    >
      <div>
        <p
          class="p-2 border border-neutral-300"
        >
          Option One text
        </p>
        <button
          class="bg-green-600 hover:bg-green-700 w-full block text-white font-bold cursor-pointer disabled:bg-neutral-400 disabled:cursor-auto "
        >
          Click
        </button>
      </div>
      <div>
        <p
          class="p-2 border border-neutral-300"
        >
          Option Two text
        </p>
        <button
          class="bg-green-600 hover:bg-green-700 w-full block text-white font-bold cursor-pointer disabled:bg-neutral-400 disabled:cursor-auto "
        >
          Click
        </button>
      </div>
    </div>
    <div
      class="space-y-4"
    >
      <div>
        <p
          class="text-lg font-semibold"
          data-testid="total-option-one"
        >
          Option One has: 2 votes (40.00% )
        </p>
      </div>
      <div>
        <p
          class="text-lg font-semibold"
          data-testid="total-option-two"
        >
          Option Two has: 3 votes (60.00% )
        </p>
      </div>
    </div>
    <div>
      <div
        class="flex gap-4 items-center"
      >
        <div
          class="space-y-2"
        >
          <div>
            <div
              class="bg-blue-500 w-4 h-4"
            />
             
            <p>
              Option One
            </p>
          </div>
          <div>
            <div
              class="bg-red-500 w-4 h-4"
            />
             
            <p>
              Option Two
            </p>
          </div>
        </div>
        <div
          class="flex justify-center items-center"
        >
          <svg
            height="120"
            width="120"
          >
            <circle
              class="text-gray-300"
              cx="60"
              cy="60"
              fill="transparent"
              r="55"
              stroke="currentColor"
              stroke-width="10"
            />
            <circle
              class="text-blue-500"
              cx="60"
              cy="60"
              fill="transparent"
              r="55"
              stroke="currentColor"
              stroke-dasharray="138.2300767579509 345.57519189487726"
              stroke-dashoffset="0"
              stroke-width="10"
              transform="rotate(-90 60 60)"
            />
            <circle
              class="text-red-500"
              cx="60"
              cy="60"
              fill="transparent"
              r="55"
              stroke="currentColor"
              stroke-dasharray="207.34511513692635 345.57519189487726"
              stroke-dashoffset="0"
              stroke-width="10"
              transform="rotate(54 60 60)"
            />
          </svg>
        </div>
      </div>
    </div>
  </section>
  ,
</DocumentFragment>
`;
